ppa/archive/views.py

```diff
@@ -249,20 +256,59 @@ class DigitizedWorkDetailView(AjaxTemplateMixin, SolrLastModifiedMixin, DetailVi
     slug_url_kwarg = "source_id"
     form_class = SearchWithinWorkForm
     paginate_by = 50
+    # see other redirect url that may be set in some cases
+    see_other_url = None
 
     def get_template_names(self):
         if self.object.status == DigitizedWork.SUPPRESSED:
             return "410.html"
         return super().get_template_names()
 
     def get_solr_lastmodified_filters(self):
-        return {"source_id": self.object.source_id}
+        if hasattr(self, "object"):
+            return {"id": self.object.index_id()}
+        # solr last modified mixin requires a filter; we don't want to return a header
+        # if we don't have an object, so return a bogus id
+        return {"id": "NOTFOUND"}
+
+    def get_queryset(self):
+        # get default queryset and filter by source id
+        source_qs = (
+            super().get_queryset().filter(source_id=self.kwargs.get("source_id"))
+        )
+        start_page = self.kwargs.get("start_page")
+        # if start page is specified, filter to get the correct excerpt
+        if start_page:
+            qs = source_qs.filter(pages_digital__startswith=start_page)
+        # if start page is NOT specified, ensure we do not retrieve an excerpt
+        else:
+            qs = source_qs.filter(pages_digital__exact="")
+
+        #  if qs is empty and start page is not set, check if there is _one_ excerpt
+        # for the source id; if there is, we want to redirect with status 303 See Other
+        if not qs.exists() and not start_page:
+            if source_qs.count() == 1:
+                self.see_other_url = source_qs.first().get_absolute_url()
+
+            # otherwise, return a 404
+
+        return qs
 
     def get(self, *args, **kwargs):
-        response = super().get(*args, **kwargs)
+        try:
+            response = super().get(*args, **kwargs)
+        except Http404:
+            # if see other url is set (i.e., tried to retrieve a non-existent
+            # full work, but there is one excerpt with that source id)
+            if self.see_other_url:
+                return HttpResponseSeeOther(self.see_other_url)
+            # otherwise, let the 404 propagate
+            raise
```

Comment:
this is a strange situation...I don't know if 303 makes sense here, but I'm also not sure anything else does. presuming we're implementing this mostly so that any links generated prior to excerpt support which appeared to point to full works will now do the right thing and point to the excerpt. [based on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/303), it looks like 303 is usually used in response to POST/PUT to indicate that the server is responding with a "you successfully uploaded X" message rather than the object you actually uploaded, which isn't our use case at all. [the actual spec](https://datatracker.ietf.org/doc/html/rfc7231#section-6.4.4) is a lot more vague, but it also only really mentions that use case.

would it be a problem to just permanently redirect with a 301? it seems like if we are including excerpts of an item in the PPA, there's no scenario in which we would also include the entire item, so it's fine to permanently redirect to the single excerpt (until such time as another excerpt is added, at which point the request is too ambiguous to resolve, so we just 404).

Reply:
very tangential: if we do use 303 and are happy with the semantics, maybe we should also use [HTTP 300](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/300) for the case where there are multiple excerpts!

Reply:
Hmm, yeah. I was thinking it shouldn't be a 301 because it isn't the same resource. I _think_ any client that doesn't care about the specifics will just see it as a redirect; but maybe the 301 is closer to what we want, if what we care about most is bookmarks and indexing.  I don't know what search engines do with 303s — [this article](https://www.seoclarity.net/pillar/url-redirects) indicates it is treated like a temporary redirect; I see other posts and discussions recommending not to use it for SEO reasons.

I considered 300, actually! From what I read, there's no standard way to generate the choice list so we'd have to generate an html page — and I wasn't sure it was worth that much effort. (Although now that I think more about it, maybe it wouldn't be that hard; maybe we could adapt our error page template and just pass in a list of choices... but still think this is a nice-to-have and we're tight on time)